### PR TITLE
[RF] Enable dirty flag propagation in `RooAbsReal::computeBatch` and speed up RooFitDriver by getting output sizes of nodes only once

### PR DESCRIPTION
**Pull request by guitargeek**
Dirty flag propagation appeared to be unnecessary in the batch mode
because the RooFitDriver manages the evaluation of nodes, but some nodes
have internal computation graphs that rely on the dirty flag
propagation, notably the integrals of a RooAbsAnaConvPdf.

This fixes some stressRooFit tests wit the RooFit batchmode:
```
Test 21 : Conditional use of per-event error p.d.f. F(t|dt)......OK
Test 28 : P.d.f. marginalization through integration.............OK
```

The only tests that still needs to be fixed now is:
```
Test 22 : Full per-event error p.d.f. F(t|dt)G(dt)...............FAILED
```

### CODE REVIEW

File: roofit/roofitcore/src/RooFitDriver.cxx [hunk 1]
```diff
@@ -359,9 +370,11 @@ void RooFitDriver::computeCPUNode(const RooAbsArg *node, NodeInfo &info)
 /// Returns the value of the top node in the computation graph
 double RooFitDriver::getVal()
 {
+   ++_getValInvocations;
+
```
Comment (lmoneta) [inline roofit/roofitcore/src/RooFitDriver.cxx:374]: I think you mean here first 3 nll evaluations

**Review by lmoneta — APPROVED** [review]
LGTM !

**Comment by lmoneta** [comment]
Thanks for this Jonas. It looks good but before merging we need too undrrstand the test failures

File: roofit/roofitcore/src/RooFitDriver.cxx (hunk 1 above)
Comment (guitargeek) [reply roofit/roofitcore/src/RooFitDriver.cxx]: Right, but it doesn't matter now because I could revert that change anyway.

**Comment by guitargeek** [comment]
Hi @lmoneta, I realized was the problem was.

What I did so far in this PR was to set all nodes to `ADirty` in the RooFitDriver constructor, and then in the first evaluation round I set back the nodes evaluated in scalar node back to `Auto`. That caused trouble with the dirty flag propagation, because clients that are also evaluated in scalar mode are only set to `Auto` later when it's their turn.

To fix this, I now figure out which nodes are evaluated in scalar mode in the RooFitDriver constructor, setting the operation mode correctly once and for all. Note that this required to cherry-pick the first commit from https://github.com/root-project/root/pull/9423 that is necessary to figure out the node output size in the RooFitDriver constructor.

**Pull request merged**